### PR TITLE
experiment: Add post extension install banner

### PR DESCRIPTION
**Pull request by AmarTrebinjac**
## Changes

Added new experimental auth banner that encourages users to install the extension if they do not already have it. I made a component for it to not have too much conditional CSS. Same goes for the "getExtensionButton" - there was just way too much fighting with the css in "Button"

Since this is only being shown on post page, and we are potentially going to run two experiments on the banner at or around the same time, I also created the PostAuthBanner component to not clutter the post page itself with too much logic surrounding the banner.

I also made slight modifications to OnboardingHeadline so that we pass it the margin we want, and not have to fight with importants as it will be reused.

preview
image

## Events

Did you introduce any new tracking events?

## Experiment

Yes, `post_banner_extension_prompt`

## Manual Testing

> [!CAUTION]
> Please make sure existing components are not breaking/affected by this PR

MI-612 #done

### Preview domain
https://mi-612.preview.app.daily.dev

### CODE REVIEW

File: packages/shared/src/components/buttons/GetExtensionButton.tsx
Comment (AmarTrebinjac) [inline packages/shared/src/components/buttons/GetExtensionButton.tsx:23]: For some reason "btn-primary" makes the chrome icon transparent, so setting the color to white solves it.

File: packages/shared/src/components/auth/AuthExtensionBanner.tsx
Comment (sshanzel) [inline packages/shared/src/components/auth/AuthExtensionBanner.tsx:11]: For the arbitrary value, we can do an inline `style` prop on the component. I think that is much easier to read. Non blocking though.

File: packages/shared/src/components/buttons/GetExtensionButton.tsx
Comment (sshanzel) [inline packages/shared/src/components/buttons/GetExtensionButton.tsx:14]: I think it might still be better to use the existing Button component, then just add the `!text-white` to fix the issue.

File: packages/webapp/pages/onboarding.tsx
Comment (sshanzel) [inline packages/webapp/pages/onboarding.tsx:328]: Was this intentional?

File: packages/webapp/pages/posts/[id]/index.tsx
Comment (sshanzel) [inline packages/webapp/pages/posts/[id]/index.tsx:69]: Just to follow the current convention 👀 
```suggestion
    /* webpackChunkName: "postAuthBanner" */ '@dailydotdev/shared/src/components/auth/PostAuthBanner'
```

**Review by sshanzel — COMMENTED** [review]
Looking great so far! Just some few suggestions and questions, though we are almost there 🚀

File: packages/webapp/pages/onboarding.tsx
Comment (AmarTrebinjac) [reply packages/webapp/pages/onboarding.tsx:328]: @sshanzel Yes, mentioned in PR description 😅 

`
I also made slight modifications to OnboardingHeadline so that we pass it the margin we want, and not have to fight with importants as it will be reused.
`

File: packages/webapp/pages/posts/[id]/index.tsx
Comment (AmarTrebinjac) [reply packages/webapp/pages/posts/[id]/index.tsx:173]: Good catch! That is definitely a mistake. Good catch!

Re-adding

File: packages/shared/src/components/buttons/GetExtensionButton.tsx
Comment (AmarTrebinjac) [reply packages/shared/src/components/buttons/GetExtensionButton.tsx:14]: It's not just the text-white, there's a bunch of class overrides, and quite a few importants (height, font, padding, etc), at which point I think it's better to just make a separate component 😛

File: packages/webapp/pages/onboarding.tsx
Comment (sshanzel) [reply packages/webapp/pages/onboarding.tsx:328]: Oh, totally missed that 😆

File: packages/shared/src/components/banners/common.ts
Comment (ilasw) [inline packages/shared/src/components/banners/common.ts:12]: Maybe is better to use image (position absolute + low z-index) for this, enabling lazy loading and srcset (this image is HUGE)

File: packages/shared/src/components/auth/AuthExtensionBanner.tsx
Comment (ilasw) [inline packages/shared/src/components/auth/AuthExtensionBanner.tsx:17]: We are not adding dynamic classes, imo we can remove `classNames`

**Review by ilasw — CHANGES_REQUESTED** [review]
Checked code wise, just some minor comments

File: packages/shared/src/components/banners/common.ts
Comment (AmarTrebinjac) [reply packages/shared/src/components/banners/common.ts:12]: @ilasw I don't know if I'm getting a little confused here or not, but 9.9kb is surely not huge for an image that spans the width of the screen?

Comment (ilasw) [reply packages/shared/src/components/banners/common.ts:12]: This image is ~20kb before compression and can be ~10kb [after resizing at 1920px width](https://res.cloudinary.com/daily-now/image/upload/s--N2EPbaJl--/c_fit,w_1920/login-popover-dailydev_mxb7lw) (this is a good size for a good part of monitors). 

image

I think that is important to try have the smallest impact on page load if we can do this with low effort. Also we already gained lot of LCP using this hack in onboarding, so this is not a new pattern. wdyt?

Comment (AmarTrebinjac) [reply packages/shared/src/components/banners/common.ts:12]: I'm a little confused. Can you explain where you are getting the 20KB number from? 
It's 10kb for me when looking in element, network tab and when downloading it

Styles:
Screenshot 2024-11-20 at 16 40 31

Network:
Screenshot 2024-11-20 at 16 38 05

Download:
Screenshot 2024-11-20 at 16 38 52

Comment (ilasw) [reply packages/shared/src/components/banners/common.ts:12]: This is the result [using the cropped image](https://res.cloudinary.com/daily-now/image/upload/s--qPSnoHHa--/c_auto,g_center,w_1920/f_auto/login-popover-dailydev_mxb7lw). Your call if implement the srcset but I strongly suggest to use it

image

Comment (AmarTrebinjac) [reply packages/shared/src/components/banners/common.ts]: @ilasw  Added it.

Thanks for uploading the sizes to cloudinary

File: packages/shared/src/components/auth/AuthExtensionBanner.tsx
Comment (rebelchris) [inline packages/shared/src/components/auth/AuthExtensionBanner.tsx:22]: When using srcSet we need to manually add onError as now the default won't trigger.

Comment (rebelchris) [reply packages/shared/src/components/auth/AuthExtensionBanner.tsx:22]: Still personally think f_auto is enough for these edge-cases though.

File: packages/shared/src/components/auth/AuthenticationBanner.tsx
Comment (rebelchris) [inline packages/shared/src/components/auth/AuthenticationBanner.tsx:16]: The duplication is also not needed, but again wouldn't even do this transform I don't think.

File: packages/shared/src/components/auth/PostAuthBanner.tsx
Comment (rebelchris) [inline packages/shared/src/components/auth/PostAuthBanner.tsx:8]: I almost think https://storybook.daily.dev/?path=/docs/experiments-helpers--docs#withexperiment-hoc is better here, but not blocking.

File: packages/shared/src/components/banners/common.ts
Comment (rebelchris) [inline packages/shared/src/components/banners/common.ts:4]: This will affect all banner cases?
Maybe it breaks something with announcement banner?

File: packages/shared/src/components/buttons/GetExtensionButton.tsx
Comment (rebelchris) [inline packages/shared/src/components/buttons/GetExtensionButton.tsx:15]: This should be a set URL somewhere I think

File: packages/shared/src/components/buttons/GetExtensionButton.tsx
Comment (rebelchris) [reply packages/shared/src/components/buttons/GetExtensionButton.tsx:14]: I guess it's non guidelined the whole thing.
I did think we had it somewhere, but can't find it now maybe it got removed before.

File: packages/shared/src/components/auth/PostAuthBanner.tsx
Comment (rebelchris) [inline packages/shared/src/components/auth/PostAuthBanner.tsx:8]: Not sure what but this now introduces a change in prod value already:
[existing-change.webm](https://github.com/user-attachments/assets/d48cdab3-f132-4449-877c-d0b0b89de5ab)

**Review by rebelchris — CHANGES_REQUESTED** [review]
Nice work, just check my comments on existing one atleast.

File: packages/shared/src/components/auth/AuthExtensionBanner.tsx
Comment (AmarTrebinjac) [reply packages/shared/src/components/auth/AuthExtensionBanner.tsx:22]: @rebelchris Ah, thanks for the bit about onError, will add.

As far as the crops go, initially I agreed, but @ilasw showed [here](https://github.com/dailydotdev/apps/pull/3847#discussion_r1850657413) that the difference was quite noticable. 1440 and 1920 save about 7 and 8kb respectively.

Although, it would've perhaps been okay to only go with the 1920 version of the image and just have that as a background.

File: packages/shared/src/components/banners/common.ts
Comment (AmarTrebinjac) [reply packages/shared/src/components/banners/common.ts:4]: Seems to only be referenced in the two banners altered in this commit, so I think we should be good

Screenshot 2024-11-21 at 10 46 23

File: packages/shared/src/components/auth/PostAuthBanner.tsx
Comment (AmarTrebinjac) [reply packages/shared/src/components/auth/PostAuthBanner.tsx:8]: @rebelchris I'm not entirely sure I follow ? 😅 You are swapping between prod and preview in that video. Both look as they are supposed to, or ?

If you're referring to the fact that the design changes, we clarified with the design team that the design changes were general changes, and only the CTA is the actual experiment:

[reference](https://dailydotdev.slack.com/archives/C0381BQUEN9/p1732103856077919?thread_ts=1731933888.885839&cid=C0381BQUEN9)

Comment (rebelchris) [reply packages/shared/src/components/auth/PostAuthBanner.tsx:8]: Ah thanks, that kind of stuff is good to note, else others might keep saying you introduced a change that's unintentional :)

File: packages/shared/src/components/auth/AuthExtensionBanner.tsx
Comment (rebelchris) [reply packages/shared/src/components/auth/AuthExtensionBanner.tsx:22]: Yeah just feels like it adds so much overhead and saving for very little endpoints.
But not really blocking

File: packages/shared/src/components/banners/common.ts
Comment (rebelchris) [reply packages/shared/src/components/banners/common.ts:4]: Cool cool

File: packages/shared/src/components/auth/AuthExtensionBanner.tsx
Comment (ilasw) [reply packages/shared/src/components/auth/AuthExtensionBanner.tsx:22]: > When using srcSet we need to manually add onError as now the default won't trigger.

Why do we need the onError? Here we don't have the change to have undefined/wrong links for this link 👀

File: packages/shared/src/components/auth/PostAuthBanner.tsx
Comment (AmarTrebinjac) [reply packages/shared/src/components/auth/PostAuthBanner.tsx:8]: Fair, I'll make sure to note these kind of things in the initial PR opening!

File: packages/shared/src/components/auth/AuthExtensionBanner.tsx
Comment (rebelchris) [reply packages/shared/src/components/auth/AuthExtensionBanner.tsx:22]: Yeah I guess these are defined by us.
So guess it's ok for this case maybe

File: packages/shared/src/components/buttons/GetExtensionButton.tsx
Comment (sshanzel) [reply packages/shared/src/components/buttons/GetExtensionButton.tsx:15]: I see. In that case, we can maybe extract this component and name it `InstallExtensionButton` - just so we won't have to create another one again.

**Review by sshanzel — APPROVED** [review]
Nothing else to add from my side! Looking great! 🚀 

Once you get approvals from the ones requesting changes, we should be able to merge :shipit:

File: packages/shared/src/components/auth/AuthExtensionBanner.tsx
Comment (ilasw) [inline packages/shared/src/components/auth/AuthExtensionBanner.tsx:12]: Why don't are we using an object for this? Maybe is better to not replace it dynamically every time

```ts
const bannerBg = { default: '', laptop: '', desktop: '' };
```

Comment (capJavert) [reply packages/shared/src/components/auth/AuthExtensionBanner.tsx:12]: I would just introduce different consts in `src/lib/image.ts`

Comment (capJavert) [reply packages/shared/src/components/auth/AuthExtensionBanner.tsx:12]: replacing can break fast if links change for any reason

**Review by ilasw — APPROVED** [review]
Looks good to me ✔️

**Pull request merged**
